qiskit/quantum_info/operators/symplectic/clifford_circuits.py

```diff
@@ -103,19 +130,75 @@ def _append_operation(clifford, operation, qargs=None):
         clifford.tableau = composed_clifford.tableau
         return clifford
 
-    # If not a Clifford basis gate we try to unroll the gate and
-    # raise an exception if unrolling reaches a non-Clifford gate.
-    # TODO: We could also check u3 params to see if they
-    # are a single qubit Clifford gate rather than raise an exception.
-    if gate.definition is None:
-        raise QiskitError(f"Cannot apply Instruction: {gate.name}")
-
-    return _append_circuit(clifford, gate.definition, qargs)
+    # If the gate is not directly appendable, we try to unroll the gate with its definition.
+    # This succeeds only if the gate has all-Clifford definition (decomposition).
+    # If fails, we need to restore the clifford that was before attempting to unroll and append.
+    if gate.definition is not None:
+        if recursion_depth > 0:
+            return _append_circuit(clifford, gate.definition, qargs, recursion_depth + 1)
+        else:  # recursion_depth == 0
+            # clifford may be updated in _append_circuit
+            org_clifford = copy.deepcopy(clifford)
+            try:
+                return _append_circuit(clifford, gate.definition, qargs, 1)
+            except (QiskitError, RecursionError):
+                # discard incompletely updated clifford and continue
+                clifford = org_clifford
+
+    # As a final attempt, if the gate is up to 3 qubits,
+    # we try to construct a Clifford to be appended from its matrix representation.
+    if isinstance(gate, Gate) and len(qargs) <= 3:
+        try:
+            matrix = gate.to_matrix()
+            try:
+                gate_cliff = Clifford.from_matrix(matrix)
+                return _append_operation(clifford, gate_cliff, qargs=qargs)
+            except QiskitError as err:
+                raise QiskitError(f"Cannot apply non-Clifford gate: {gate.name}") from err
+        except TypeError as err:
+            raise QiskitError(f"Cannot apply {gate.name} gate with unbounded parameters") from err
+        except CircuitError:
+            pass
+
+    raise QiskitError(f"Cannot apply {gate}")
+
+
+def _n_half_pis(param) -> int:
+    try:
```

Comment:
How about `if not isinstance(param, ParameterExpression)`?

Reply:
We need to try cast here because that checks if a ParameterExpression is *bounded or not* (not checking the type of `param`).